File: scripts/sync_markdown.py
#!/usr/bin/env python3
"""Sync changed markdown files and .alternates mappings to archive.tw.

Invoked by .github/workflows/upload-markdown-on-change.yml. All state comes from
environment variables (API_ENDPOINT, TOKEN, BEFORE_SHA, AFTER_SHA, and the
GitHub-provided GITHUB_WORKSPACE / GITHUB_EVENT_NAME / GITHUB_REF_NAME).

Pass --dry-run to print every intended HTTP request without actually sending it.
"""

import json
import os
import pathlib
import subprocess
import sys
import time
import urllib.error
import urllib.parse
import urllib.request

DRY_RUN = "--dry-run" in sys.argv

endpoint = os.environ["API_ENDPOINT"]
token = os.environ["TOKEN"]
workspace = pathlib.Path(os.environ["GITHUB_WORKSPACE"])
before_sha = os.environ.get("BEFORE_SHA", "")
after_sha = os.environ.get("AFTER_SHA", "")
event_name = os.environ.get("GITHUB_EVENT_NAME", "")
ref_name = os.environ.get("GITHUB_REF_NAME", "")

def parse_alternates_text(raw_text):
  filename_pairs = {}
  slug_pairs = {}
  if not raw_text:
    return filename_pairs, slug_pairs

  for line in raw_text.splitlines():
    line = line.strip()
    if not line or line.startswith("#"):
      continue
    parts = line.split("\t")
    if len(parts) == 2:
      a = pathlib.Path(parts[0].strip()).name
      b = pathlib.Path(parts[1].strip()).name
      if pathlib.Path(a).suffix != ".md" or pathlib.Path(b).suffix != ".md":
        continue
      filename_pairs[a] = b
      filename_pairs[b] = a
      slug_pairs[a] = pathlib.Path(b).stem.lower()
      slug_pairs[b] = pathlib.Path(a).stem.lower()
  return filename_pairs, slug_pairs

def read_file_at_ref(ref, rel_path):
  if not ref or ref == ("0" * 40):
    return None
  try:
    return subprocess.check_output(
      ["git", "show", f"{ref}:{rel_path}"],
      encoding="utf-8",
      errors="replace",
    )
  except subprocess.CalledProcessError:
    return None

# Parse current .alternates file: bidirectional map of filename → alternate slug
alternates = {}
current_alt_pairs = {}
alt_path = workspace / ".alternates"
if alt_path.exists():
  current_alt_pairs, alternates = parse_alternates_text(
    alt_path.read_text(encoding="utf-8")
  )
  print(f"[debug] loaded {len(alternates)} alternate mappings")

MAX_RETRIES = 3
BASE_DELAY = 2  # seconds

def request_with_retry(req, label, max_retries=MAX_RETRIES):
  """Send an HTTP request with exponential backoff retry on 5xx errors."""
  if DRY_RUN:
    payload_preview = ""
    if req.data:
      try:
        payload_preview = json.dumps(json.loads(req.data), ensure_ascii=False, indent=2)
      except (ValueError, TypeError):
        payload_preview = repr(req.data)[:500]
    print(f"[dry-run] {req.get_method()} {req.full_url}")
    if payload_preview:
      print(f"[dry-run]   payload: {payload_preview}")
    print(f"{label} -> HTTP 200 (dry-run)")
    return 200, "[dry-run]"
  for attempt in range(1, max_retries + 1):
    try:
      with urllib.request.urlopen(req) as resp:
        body = resp.read().decode("utf-8", errors="replace")
        print(f"{label} -> HTTP {resp.status}")
        print(body)
        return resp.status, body
    except urllib.error.HTTPError as e:
      body = e.read().decode("utf-8", errors="replace")
      if e.code >= 500 and attempt < max_retries:
        delay = BASE_DELAY * (2 ** (attempt - 1))
        print(f"{label} failed -> HTTP {e.code} (attempt {attempt}/{max_retries}, retrying in {delay}s)")
        print(body)
        time.sleep(delay)
        # Re-create the request (the previous one's data stream is consumed)
        new_req = urllib.request.Request(
          req.full_url,
          data=req.data,
          method=req.get_method(),
        )
        for header, value in req.headers.items():
          new_req.add_header(header, value)
        req = new_req
        continue
      print(f"{label} failed -> HTTP {e.code}")
      print(body)
      raise

print(f"[debug] event={event_name} ref={ref_name}")
print(f"[debug] before={before_sha}")
print(f"[debug] after={after_sha}")
if DRY_RUN:
  print("[debug] DRY-RUN mode: no HTTP requests will be sent")

if before_sha and before_sha != ("0" * 40):
  cmd = ["git", "diff", "--name-status", "--no-renames", "-z", before_sha, after_sha]
else:
  # Fallback for the first push where "before" can be all zeros.
  cmd = ["git", "show", "--name-status", "--pretty=format:", "--no-renames", "-z", after_sha]

print(f"[debug] cmd={' '.join(cmd)}")
diff_output = subprocess.check_output(cmd, encoding="utf-8")
print("[debug] raw diff output:")
print(repr(diff_output) if diff_output else "(empty)")

added_files = []
modified_files = []
removed_files = []
alternates_changed = False

# Parse NUL-delimited output: format is <status> NUL <path> NUL [ <status> NUL <path> NUL ... ]
tokens = [t.strip() for t in diff_output.split("\0") if t.strip()]
i = 0
while i + 1 < len(tokens):
  status = tokens[i]
  rel_path = tokens[i + 1]
  i += 2
  # Defensive: unquote if needed
  if rel_path.startswith('"') and rel_path.endswith('"') and len(rel_path) >= 2:
    rel_path = rel_path[1:-1]
  p = pathlib.Path(rel_path)

  if rel_path == ".alternates":
    alternates_changed = True
    print(f"[debug] detect .alternates change: status={status!r}")
    continue

  # 只處理根目錄 markdown 檔
  if p.suffix != ".md" or len(p.parts) != 1:
    print(f"[debug] skip by path filter: status={status!r} path={rel_path!r}")
    continue

  if status.startswith("A"):
    added_files.append(rel_path)
    print(f"[debug] detect added: {rel_path!r}")
  elif status.startswith("M"):
    modified_files.append(rel_path)
    print(f"[debug] detect modified: {rel_path!r}")
  elif status.startswith("D"):
    removed_files.append(rel_path)
    print(f"[debug] detect removed: {rel_path!r}")

# Keep order while removing duplicates.
unique_added_files = list(dict.fromkeys(added_files))
unique_modified_files = list(dict.fromkeys(modified_files))
unique_removed_files = list(dict.fromkeys(removed_files))
print(f"[debug] added root md files={unique_added_files}")
print(f"[debug] modified root md files={unique_modified_files}")
print(f"[debug] removed root md files={unique_removed_files}")

if alternates_changed:
  before_alt_pairs, _ = parse_alternates_text(
    read_file_at_ref(before_sha, ".alternates")
  )
  affected_alt_files = sorted(
    rel_path
    for rel_path in (set(before_alt_pairs) | set(current_alt_pairs))
    if before_alt_pairs.get(rel_path) != current_alt_pairs.get(rel_path)
  )
  print(f"[debug] .alternates affected root md files={affected_alt_files}")

  for rel_path in affected_alt_files:
    if rel_path in unique_added_files or rel_path in unique_modified_files or rel_path in unique_removed_files:
      continue

    p = pathlib.Path(rel_path)
    if p.suffix != ".md" or len(p.parts) != 1:
      print(f"[debug] skip non-root alternate target: {rel_path!r}")
      continue

    abs_path = workspace / rel_path
    if abs_path.exists():
      unique_modified_files.append(rel_path)
      print(f"[debug] re-sync due to .alternates change: {rel_path!r}")
    else:
      unique_removed_files.append(rel_path)
      print(f"[debug] delete stale alternate target: {rel_path!r}")

if not unique_added_files and not unique_modified_files and not unique_removed_files:
  print("No root-level markdown or .alternates changes requiring sync.")
  sys.exit(0)

failures = []

for rel_path in unique_added_files:
  abs_path = workspace / rel_path
  if not abs_path.exists():
    print(f"Skip missing file: {rel_path}")
    continue

  try:
    markdown = abs_path.read_text(encoding="utf-8")
    payload = {
      "filename": rel_path,
      "markdown": markdown,
    }
    if rel_path in alternates:
      payload["alternate_filename"] = alternates[rel_path]
      print(f"[debug] alternate for {rel_path}: {alternates[rel_path]}")
    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
      endpoint,
      data=data,
      method="POST",
      headers={
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json; charset=utf-8",
        "User-Agent": "github-actions/1.0",
        "X-GitHub-Repository": "audreyt/transcript",
      },
    )

    request_with_retry(req, f"POST {rel_path}")
  except Exception as e:
    print(f"ERROR posting {rel_path}: {e}")
    failures.append(rel_path)

for rel_path in unique_modified_files:
  abs_path = workspace / rel_path
  if not abs_path.exists():
    print(f"Skip missing file: {rel_path}")
    continue

  try:
    markdown = abs_path.read_text(encoding="utf-8")
    payload = {
      "filename": rel_path,
      "markdown": markdown,
      "alternate_filename": alternates.get(rel_path),
    }
    print(f"[debug] alternate for {rel_path}: {payload['alternate_filename']}")
    data = json.dumps(payload).encode("utf-8")
    headers = {
      "Authorization": f"Bearer {token}",
      "Content-Type": "application/json; charset=utf-8",
      "User-Agent": "github-actions/1.0",
      "X-GitHub-Repository": "audreyt/transcript",
    }
    req = urllib.request.Request(endpoint, data=data, method="PATCH", headers=headers)
    try:
      request_with_retry(req, f"PATCH {rel_path}")
    except urllib.error.HTTPError as patch_err:
      if patch_err.code == 404:
        print(f"PATCH {rel_path} got 404, falling back to POST")
        post_data = json.dumps(payload).encode("utf-8")
        post_req = urllib.request.Request(endpoint, data=post_data, method="POST", headers=headers)
        request_with_retry(post_req, f"POST (fallback) {rel_path}")
      else:
        raise
  except Exception as e:
    print(f"ERROR syncing {rel_path}: {e}")
    failures.append(rel_path)

for rel_path in unique_removed_files:
  # filename を query param で送る（CF が DELETE body を剥ぎ取るため）
  delete_url = endpoint + "?filename=" + urllib.parse.quote(rel_path, safe="")
  req = urllib.request.Request(
    delete_url,
    method="DELETE",
    headers={
      "Authorization": f"Bearer {token}",
      "User-Agent": "github-actions/1.0",
      "X-GitHub-Repository": "audreyt/transcript",
    },
  )

  try:
    request_with_retry(req, f"DELETE {rel_path}")
  except urllib.error.HTTPError as e:
    if e.code == 404:
      print(f"DELETE {rel_path}: already absent (HTTP 404), skipping")
    else:
      print(f"ERROR deleting {rel_path}: {e}")
      failures.append(rel_path)
  except Exception as e:
    print(f"ERROR deleting {rel_path}: {e}")
    failures.append(rel_path)

if failures:
  print(f"\nFailed to sync {len(failures)} file(s): {failures}")
  sys.exit(1)

File: scripts/sync_markdown.ts
#!/usr/bin/env bun
/**
 * Sync changed markdown files and .alternates mappings to archive.tw.
 * Also syncs the .redirects snapshot (speech_redirects table) on every run.
 *
 * Environment variables:
 *   API_ENDPOINT             — POST/PATCH/DELETE markdown
 *   API_REDIRECTS_ENDPOINT   — PUT .redirects snapshot (optional; skipped if unset)
 *   TOKEN
 *   GITHUB_WORKSPACE
 *   BEFORE_SHA
 *   AFTER_SHA
 *   GITHUB_EVENT_NAME
 *   GITHUB_REF_NAME
 *
 * Pass --dry-run to print intended HTTP requests without sending them.
 */

import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { execFileSync } from "node:child_process";

export const MAX_RETRIES = 3;
export const BASE_DELAY_MS = 2_000;

export interface SyncEnvironment {
  API_ENDPOINT: string;
  API_REDIRECTS_ENDPOINT?: string;
  TOKEN: string;
  GITHUB_WORKSPACE: string;
  BEFORE_SHA?: string;
  AFTER_SHA?: string;
  GITHUB_EVENT_NAME?: string;
  GITHUB_REF_NAME?: string;
}

export interface SyncDeps {
  git?: (args: string[], cwd: string) => string;
  fetchImpl?: typeof fetch;
  sleep?: (ms: number) => Promise<void>;
  stdout?: (line: string) => void;
}

export interface RequestSpec {
  url: string;
  method: string;
  headers?: Record<string, string>;
  body?: string;
}

export interface AlternatesParseResult {
  filenamePairs: Record<string, string>;
  slugPairs: Record<string, string>;
}

export interface DiffSummary {
  addedFiles: string[];
  modifiedFiles: string[];
  removedFiles: string[];
  alternatesChanged: boolean;
}

function defaultGit(args: string[], cwd: string): string {
  return execFileSync("git", args, {
    cwd,
    encoding: "utf-8",
    stdio: ["ignore", "pipe", "pipe"],
  });
}

function defaultSleep(ms: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, ms);
  });
}

export function parseAlternatesText(rawText: string | null | undefined): AlternatesParseResult {
  const filenamePairs: Record<string, string> = {};
  const slugPairs: Record<string, string> = {};
  if (!rawText) {
    return { filenamePairs, slugPairs };
  }

  for (const rawLine of rawText.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) {
      continue;
    }
    const parts = line.split("\t");
    if (parts.length !== 2) {
      continue;
    }
    const a = path.basename(parts[0].trim());
    const b = path.basename(parts[1].trim());
    if (!a.endsWith(".md") || !b.endsWith(".md")) {
      continue;
    }

    filenamePairs[a] = b;
    filenamePairs[b] = a;
    slugPairs[a] = path.parse(b).name.toLowerCase();
    slugPairs[b] = path.parse(a).name.toLowerCase();
  }

  return { filenamePairs, slugPairs };
}

export function readFileAtRef(
  ref: string | undefined,
  relativePath: string,
  workspace: string,
  git: (args: string[], cwd: string) => string = defaultGit,
): string | null {
  if (!ref || ref === "0".repeat(40)) {
    return null;
  }

  try {
    return git(["show", `${ref}:${relativePath}`], workspace);
  } catch {
    return null;
  }
}

function uniq(values: string[]): string[] {
  return [...new Set(values)];
}

function isRootMarkdown(relativePath: string): boolean {
  return relativePath.endsWith(".md") && !relativePath.includes("/") && !relativePath.includes(path.sep);
}

export function parseNameStatusDiff(diffOutput: string): DiffSummary {
  const addedFiles: string[] = [];
  const modifiedFiles: string[] = [];
  const removedFiles: string[] = [];
  let alternatesChanged = false;

  const tokens = diffOutput
    .split("\0")
    .map((token) => token.trim())
    .filter(Boolean);

  for (let index = 0; index + 1 < tokens.length; index += 2) {
    const status = tokens[index];
    let relativePath = tokens[index + 1];
    if (
      relativePath.startsWith("\"") &&
      relativePath.endsWith("\"") &&
      relativePath.length >= 2
    ) {
      relativePath = relativePath.slice(1, -1);
    }

    if (relativePath === ".alternates") {
      alternatesChanged = true;
      continue;
    }

    if (!isRootMarkdown(relativePath)) {
      continue;
    }

    if (status.startsWith("A")) {
      addedFiles.push(relativePath);
    } else if (status.startsWith("M")) {
      modifiedFiles.push(relativePath);
    } else if (status.startsWith("D")) {
      removedFiles.push(relativePath);
    }
  }

  return {
    addedFiles: uniq(addedFiles),
    modifiedFiles: uniq(modifiedFiles),
    removedFiles: uniq(removedFiles),
    alternatesChanged,
  };
}

export function applyAlternatesImpact(
  summary: DiffSummary,
  beforeAltPairs: Record<string, string>,
  currentAltPairs: Record<string, string>,
  workspace: string,
): DiffSummary {
  if (!summary.alternatesChanged) {
    return summary;
  }

  const addedFiles = [...summary.addedFiles];
  const modifiedFiles = [...summary.modifiedFiles];
  const removedFiles = [...summary.removedFiles];

  const affectedFiles = uniq(
    [...Object.keys(beforeAltPairs), ...Object.keys(currentAltPairs)].filter(
      (relativePath) => beforeAltPairs[relativePath] !== currentAltPairs[relativePath],
    ),
  ).sort();

  for (const relativePath of affectedFiles) {
    if (
      addedFiles.includes(relativePath) ||
      modifiedFiles.includes(relativePath) ||
      removedFiles.includes(relativePath) ||
      !isRootMarkdown(relativePath)
    ) {
      continue;
    }

    if (existsSync(path.join(workspace, relativePath))) {
      modifiedFiles.push(relativePath);
    } else {
      removedFiles.push(relativePath);
    }
  }

  return {
    addedFiles: uniq(addedFiles),
    modifiedFiles: uniq(modifiedFiles),
    removedFiles: uniq(removedFiles),
    alternatesChanged: summary.alternatesChanged,
  };
}

export async function requestWithRetry(
  spec: RequestSpec,
  label: string,
  {
    dryRun = false,
    fetchImpl = fetch,
    sleep = defaultSleep,
    stdout = (line: string) => console.log(line),
    maxRetries = MAX_RETRIES,
  }: {
    dryRun?: boolean;
    fetchImpl?: typeof fetch;
    sleep?: (ms: number) => Promise<void>;
    stdout?: (line: string) => void;
    maxRetries?: number;
  } = {},
): Promise<{ status: number; body: string }> {
  if (dryRun) {
    stdout(`[dry-run] ${spec.method} ${spec.url}`);
    if (spec.body) {
      try {
        stdout(
          `[dry-run]   payload: ${JSON.stringify(JSON.parse(spec.body), null, 2)}`,
        );
      } catch {
        stdout(`[dry-run]   payload: ${spec.body.slice(0, 500)}`);
      }
    }
    stdout(`${label} -> HTTP 200 (dry-run)`);
    return { status: 200, body: "[dry-run]" };
  }

  for (let attempt = 1; attempt <= maxRetries; attempt += 1) {
    const response = await fetchImpl(spec.url, {
      method: spec.method,
      headers: spec.headers,
      body: spec.body,
    });
    const body = await response.text();

    if (response.status >= 500 && attempt < maxRetries) {
      const delay = BASE_DELAY_MS * 2 ** (attempt - 1);
      stdout(
        `${label} failed -> HTTP ${response.status} (attempt ${attempt}/${maxRetries}, retrying in ${delay / 1000}s)`,
      );
      stdout(body);
      await sleep(delay);
      continue;
    }

    if (!response.ok) {
      stdout(`${label} failed -> HTTP ${response.status}`);
      stdout(body);
      throw new Error(`HTTP ${response.status}: ${body}`);
    }

    stdout(`${label} -> HTTP ${response.status}`);
    stdout(body);
    return { status: response.status, body };
  }

  throw new Error(`exhausted retries for ${label}`);
}

function debug(stdout: (line: string) => void, message: string): void {
  stdout(`[debug] ${message}`);
}

function readWorkspaceFile(filePath: string): string {
  return readFileSync(filePath, "utf-8");
}

export function diffCommand(beforeSha: string, afterSha: string): string[] {
  if (beforeSha && beforeSha !== "0".repeat(40)) {
    return ["diff", "--name-status", "--no-renames", "-z", beforeSha, afterSha];
  }
  return ["show", "--name-status", "--pretty=format:", "--no-renames", "-z", afterSha];
}

function commonHeaders(token: string): Record<string, string> {
  return {
    Authorization: `Bearer ${token}`,
    "User-Agent": "github-actions/1.0",
    "X-GitHub-Repository": "audreyt/transcript",
  };
}

function buildUpsertSpec(
  endpoint: string,
  token: string,
  method: "POST" | "PATCH",
  payload: Record<string, string | null>,
): RequestSpec {
  return {
    url: endpoint,
    method,
    headers: {
      ...commonHeaders(token),
      "Content-Type": "application/json; charset=utf-8",
    },
    body: JSON.stringify(payload),
  };
}

function buildDeleteSpec(endpoint: string, token: string, relativePath: string): RequestSpec {
  const url = `${endpoint}?filename=${encodeURIComponent(relativePath)}`;
  return {
    url,
    method: "DELETE",
    headers: commonHeaders(token),
  };
}

function buildRedirectsSpec(endpoint: string, token: string, body: string): RequestSpec {
  return {
    url: endpoint,
    method: "PUT",
    headers: {
      ...commonHeaders(token),
      "Content-Type": "text/plain; charset=utf-8",
    },
    body,
  };
}

export async function runSync(
  env: SyncEnvironment,
  argv: string[] = [],
  deps: SyncDeps = {},
): Promise<number> {
  const dryRun = argv.includes("--dry-run");
  const stdout = deps.stdout ?? ((line: string) => console.log(line));
  const git = deps.git ?? defaultGit;
  const fetchImpl = deps.fetchImpl ?? fetch;
  const sleep = deps.sleep ?? defaultSleep;

  const endpoint = env.API_ENDPOINT;
  const token = env.TOKEN;
  const workspace = path.resolve(env.GITHUB_WORKSPACE);
  const beforeSha = env.BEFORE_SHA ?? "";
  const afterSha = env.AFTER_SHA ?? "";
  const eventName = env.GITHUB_EVENT_NAME ?? "";
  const refName = env.GITHUB_REF_NAME ?? "";

  if (!endpoint) {
    throw new Error("API_ENDPOINT is required");
  }
  if (!token) {
    throw new Error("TOKEN is required");
  }

  const alternatesPath = path.join(workspace, ".alternates");
  let currentAltPairs: Record<string, string> = {};
  let alternates: Record<string, string> = {};
  if (existsSync(alternatesPath)) {
    const parsed = parseAlternatesText(readWorkspaceFile(alternatesPath));
    currentAltPairs = parsed.filenamePairs;
    alternates = parsed.slugPairs;
    debug(stdout, `loaded ${Object.keys(alternates).length} alternate mappings`);
  }

  debug(stdout, `event=${eventName} ref=${refName}`);
  debug(stdout, `before=${beforeSha}`);
  debug(stdout, `after=${afterSha}`);
  if (dryRun) {
    debug(stdout, "DRY-RUN mode: no HTTP requests will be sent");
  }

  const command = diffCommand(beforeSha, afterSha);
  debug(stdout, `cmd=git ${command.join(" ")}`);
  const diffOutput = git(command, workspace);
  debug(stdout, "raw diff output:");
  stdout(diffOutput ? JSON.stringify(diffOutput) : "(empty)");

  const initialSummary = parseNameStatusDiff(diffOutput);
  debug(stdout, `added root md files=${JSON.stringify(initialSummary.addedFiles)}`);
  debug(stdout, `modified root md files=${JSON.stringify(initialSummary.modifiedFiles)}`);
  debug(stdout, `removed root md files=${JSON.stringify(initialSummary.removedFiles)}`);

  let summary = initialSummary;
  if (summary.alternatesChanged) {
    const beforeAltPairs = parseAlternatesText(
      readFileAtRef(beforeSha, ".alternates", workspace, git),
    ).filenamePairs;
    const affectedFiles = uniq(
      [...Object.keys(beforeAltPairs), ...Object.keys(currentAltPairs)].filter(
        (relativePath) => beforeAltPairs[relativePath] !== currentAltPairs[relativePath],
      ),
    ).sort();
    debug(stdout, `.alternates affected root md files=${JSON.stringify(affectedFiles)}`);
    summary = applyAlternatesImpact(summary, beforeAltPairs, currentAltPairs, workspace);
  }

  if (
    summary.addedFiles.length === 0 &&
    summary.modifiedFiles.length === 0 &&
    summary.removedFiles.length === 0
  ) {
    stdout("No root-level markdown or .alternates changes requiring sync.");
    return 0;
  }

  const failures: string[] = [];

  for (const relativePath of summary.addedFiles) {
    const absolutePath = path.join(workspace, relativePath);
    if (!existsSync(absolutePath)) {
      stdout(`Skip missing file: ${relativePath}`);
      continue;
    }

    try {
      const payload: Record<string, string | null> = {
        filename: relativePath,
        markdown: readWorkspaceFile(absolutePath),
      };
      if (alternates[relativePath]) {
        payload.alternate_filename = alternates[relativePath];
        debug(stdout, `alternate for ${relativePath}: ${alternates[relativePath]}`);
      }

      await requestWithRetry(
        buildUpsertSpec(endpoint, token, "POST", payload),
        `POST ${relativePath}`,
        { dryRun, fetchImpl, sleep, stdout },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      stdout(`ERROR posting ${relativePath}: ${message}`);
      failures.push(relativePath);
    }
  }

  for (const relativePath of summary.modifiedFiles) {
    const absolutePath = path.join(workspace, relativePath);
    if (!existsSync(absolutePath)) {
      stdout(`Skip missing file: ${relativePath}`);
      continue;
    }

    const payload: Record<string, string | null> = {
      filename: relativePath,
      markdown: readWorkspaceFile(absolutePath),
      alternate_filename: alternates[relativePath] ?? null,
    };
    if (payload.alternate_filename) {
      debug(stdout, `alternate for ${relativePath}: ${payload.alternate_filename}`);
    }

    try {
      await requestWithRetry(
        buildUpsertSpec(endpoint, token, "PATCH", payload),
        `PATCH ${relativePath}`,
        { dryRun, fetchImpl, sleep, stdout },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      if (message.startsWith("HTTP 404:")) {
        stdout(`PATCH ${relativePath} got 404, falling back to POST`);
        try {
          await requestWithRetry(
            buildUpsertSpec(endpoint, token, "POST", payload),
            `POST (fallback) ${relativePath}`,
            { dryRun, fetchImpl, sleep, stdout },
          );
        } catch (fallbackError) {
          const fallbackMessage =
            fallbackError instanceof Error ? fallbackError.message : String(fallbackError);
          stdout(`ERROR syncing ${relativePath}: ${fallbackMessage}`);
          failures.push(relativePath);
        }
      } else {
        stdout(`ERROR syncing ${relativePath}: ${message}`);
        failures.push(relativePath);
      }
    }
  }

  for (const relativePath of summary.removedFiles) {
    try {
      await requestWithRetry(
        buildDeleteSpec(endpoint, token, relativePath),
        `DELETE ${relativePath}`,
        { dryRun, fetchImpl, sleep, stdout },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      if (message.startsWith("HTTP 404:")) {
        stdout(`DELETE ${relativePath}: already absent (HTTP 404), skipping`);
      } else {
        stdout(`ERROR deleting ${relativePath}: ${message}`);
        failures.push(relativePath);
      }
    }
  }

  // Sync .redirects snapshot to /api/redirects (speech_redirects SoT).
  // Runs every time, independent of the markdown diff: even if no .md
  // changed, the .redirects file may have been edited on its own.
  const redirectsEndpoint = env.API_REDIRECTS_ENDPOINT;
  const redirectsPath = path.join(workspace, ".redirects");
  if (redirectsEndpoint && existsSync(redirectsPath)) {
    try {
      const redirectsBody = readWorkspaceFile(redirectsPath);
      await requestWithRetry(
        buildRedirectsSpec(redirectsEndpoint, token, redirectsBody),
        "PUT .redirects",
        { dryRun, fetchImpl, sleep, stdout },
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      stdout(`ERROR syncing .redirects: ${message}`);
      failures.push(".redirects");
    }
  } else if (!redirectsEndpoint) {
    debug(stdout, "API_REDIRECTS_ENDPOINT not set, skipping .redirects sync");
  } else {
    debug(stdout, ".redirects file absent, skipping redirects sync");
  }

  if (failures.length > 0) {
    stdout(`\nFailed to sync ${failures.length} file(s): ${JSON.stringify(failures)}`);
    return 1;
  }

  return 0;
}

export async function main(
  argv: string[] = Bun.argv.slice(2),
  env: NodeJS.ProcessEnv = process.env,
): Promise<number> {
  return runSync(
    {
      API_ENDPOINT: env.API_ENDPOINT ?? "",
      API_REDIRECTS_ENDPOINT: env.API_REDIRECTS_ENDPOINT,
      TOKEN: env.TOKEN ?? "",
      GITHUB_WORKSPACE: env.GITHUB_WORKSPACE ?? process.cwd(),
      BEFORE_SHA: env.BEFORE_SHA,
      AFTER_SHA: env.AFTER_SHA,
      GITHUB_EVENT_NAME: env.GITHUB_EVENT_NAME,
      GITHUB_REF_NAME: env.GITHUB_REF_NAME,
    },
    argv,
  );
}

if (import.meta.main) {
  process.exit(await main());
}
